### PR TITLE
Indirect Bayes - BayesStretch algorithm

### LINKED ISSUE #13956: Mantid Algorithm for Bayes Stretched
Bayes Stretch is currently a python script. This should be made into a Mantid Algorithm instead. The algorithm should be called BayesStretch.

### PR DESCRIPTION
**Pull request by ElliotAOram**
The functionality from the Quest function in the IndirectBayes.py script is replaced by this mantid algorithm. The interface has been updated to use this algorithm and there is an accompanying test and documentation.

The `RunQuest` function in the IndirectBayes.py script should be removed once the we have confirmed that no one is using it.
# To test

**Tests must be performed on windows due to reliance on compiled F2Py modules.**

The following script should test the functionality of the algorithm:
_`irs26176_graphite002_red` and `irs26173_graphite002_res` can be obtained from the unit test directory_

```
irs26176_graphite002_red = Load('irs26176_graphite002_red.nxs')
irs26173_graphite002_res = Load('irs26173_graphite002_res.nxs')

 # Run BayesStretch algorithm
fit_group, contour_group = BayesStretch(SampleWorkspace=irs26176_graphite002_red, ResolutionWorkspace=irs26173_graphite002_res, 
                                                        EMin=-0.2, EMax=0.2, Background='Sloping', Loop=True)

```
- Ensure this script produces 2 workspace groups 
  - The group suffixed with `_Contour` should contain 10 workspaces suffixed with `_Zp<WorkspaceIndex>`
  - The group suffixed with `_Fit` should contain the 2 workspaces (one suffixed with `_Sigma` the other `_Beta`)

Using the same sample and resolution workspaces as those defined in the script, try using the interface to perform the same operation:
- Open Stretch (Interfaces > Indirect > Bayes > Stretch)
- Add the sample (irs26176) and resolution (irs26173)
- Check the `Sequential Fit` option
- Change the `Plot Result` (combo box in the bottom left) to `All`
- Check Save Result
- Click `Run`
- Ensure the algorithm runs without error and two plots are produced (see below)
- Ensure that the workspace groups produced by the algorithm have been saved to your default save directory.

image

image

Fixes #13956 

[RELEASE NOTES](https://github.com/mantidproject/mantid/blob/a37025b5a040fdfb5ea27d0e58e3fcac3d124c44/docs/source/release/v3.8.0/indirect_inelastic.rst)

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [x] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [x] Has the relevant documentation been added/updated?
- [x] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/BayesStretch.py [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+            SaveNexusProcessed(InputWorkspace=contour_ws,
+                               Filename=contour_path)
+            logger.information('Output Contour file created : %s' % contour_path)
+            log_prog.report('Files Saved')
+
+        self.setProperty('OutputWorkspaceFit', fit_ws)
+        self.setProperty('OutputWorkspaceContour', contour_ws)
+        log_prog.report('Setting workspace properties')
+
+AlgorithmFactory.subscribe(BayesStretch)         # Register algorithm with Mantid
```
Comment (KarlPalmen) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/BayesStretch.py]: This is long. The PyExec is long.

Comment (ElliotAOram) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/BayesStretch.py]: @KarlPalmen I think you are right. I have refactored the PyExec to be shorter (and hopefully more readable). Let me know if you still think it's too long and I can give it another go.

**Comment by KarlPalmen** [comment]
There is an error in the documentation. I think it arises from inconsistent naming (BayesStretch and BayesStretched).

File: docs/source/algorithms/BayesStretched-v1.rst [hunk 2]
```diff
@@ -0,0 +1,40 @@
+.. algorithm::
```
Comment (KarlPalmen) [inline docs/source/algorithms/BayesStretched-v1.rst]: The RST file has the wrong name.

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/BayesStretch.py [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+
+    def PyInit(self):
+        self.declareProperty(MatrixWorkspaceProperty('SampleWorkspace', '', direction=Direction.Input),
+                             doc='Name of the Sample input Workspace')
+
+        self.declareProperty(MatrixWorkspaceProperty('ResolutionWorkspace', '', direction=Direction.Input),
+                             doc='Name of the resolution input Workspace')
+
+        self.declareProperty(name='MinRange', defaultValue=-0.2,
+                             doc='The start of the fit range. Default=-0.2')
```
Comment (KarlPalmen) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/BayesStretch.py]: Duplicate specification of default: Default not needed in the doc= string.

**Comment by KarlPalmen** [comment]
While testing, the Nexus file that was created, loaded OK. Documentation is now good. :shipit:

**Pull request merged**